code/src/reducers/game.js

```diff
@@ -0,0 +1,86 @@
+import { createSlice } from "@reduxjs/toolkit";
+import { ui } from "./ui";
+
+export const game = createSlice({
+  name: "game",
+  initialState: {
+    type: "move",
+    username: localStorage.getItem("username"),
+    history: [],
+    currentStep: localStorage.getItem("current-step")
+      ? JSON.parse(localStorage.getItem("current-step"))
+      : {
+          coordinates: 0.0,
+          description: "",
+          actions: [],
+        },
+  },
+  reducers: {
+    setUsername: (store, action) => {
+      store.username = action.payload;
+    },
+    setCurrentStep: (store, action) => {
+      if (store.currentStep) {
+        store.history = [...store.history, action.payload];
+      }
+      store.currentStep = action.payload;
+    },
+    setPreviousStep: (store) => {
+      if (store.history.length) {
+        store.currentStep = store.history[store.history.length - 1];
+        store.history = store.history.slice(0, store.history.length - 1);
+      }
+    },
+    restartGame: (store) => {
+      store.username = "";
+      store.currentStep = {};
+      store.history = [];
+    },
+  },
+});
+
+export const generateGame = (username) => {
+  return (dispatch) => {
```

Comment:
It would be great if you could add dispatch(ui.actions.setLoading(true)); here as well to render the loading as the game is starting. Now there's a delay without the loader 🤓